front/components/assistant_builder/AssistantBuilderTablesModal.tsx

```diff
@@ -55,45 +114,199 @@ export default function AssistantBuilderTablesModal({
+  const [parentsIntializationError, setParentsIntializationError] =
+    useState<boolean>(false);
+
+  const initializeManagedTables = useCallback(
+    async (
+      dataSource: DataSourceType,
+      tableConfigs: AssistantBuilderTableConfiguration[]
+    ) => {
+      const tableIds = tableConfigs.map((config) => config.tableId);
+
+      const contentNodes = await getContentNodes(
+        owner.sId,
+        dataSource.name,
+        tableIds
+      );
+
+      if (contentNodes.length !== tableIds.length) {
+        throw new Error(
+          `Failed to fetch content nodes for all tables. Expected ${tableIds.length}, got ${contentNodes.length}.`
+        );
+      }
+
+      setSelectedManagedTables((prev) => [
+        ...prev.filter(
+          (n) => !contentNodes.some((c) => c.internalId === n.internalId)
+        ),
+        ...contentNodes,
+      ]);
+    },
+    [owner.sId, setSelectedManagedTables]
+  );
+
+  const initializeParents = useCallback(
+    async (internalIds: string[], dataSource: DataSourceType) => {
+      const parents = await getParents(owner.sId, dataSource.name, [
+        ...internalIds,
+      ]);
+      setParentsById(parents);
+    },
+    [owner.sId]
+  );
+
+  useEffect(() => {
+    // When coming back to the modal, if the user selects a data source that
+    // is a connection and there are already tables selected, we need to
+    // fetch the content nodes for those tables and add them to the selected
+    // managed tables.
+
+    if (
+      !selectedDataSource ||
+      !selectedDataSource.connectorId ||
+      isInitializingNodes ||
+      !!selectedManagedTables.length ||
+      !Object.keys(tablesQueryConfiguration).length ||
+      !!nodeIntializationError
+    ) {
+      return;
+    }
+    const tableConfigs = Object.values(tablesQueryConfiguration).filter(
+      (c) => c.dataSourceId === selectedDataSource.name
+    );
+    if (!tableConfigs.length) {
+      return;
+    }
+    setIsInitializingNodes(true);
+    void initializeManagedTables(selectedDataSource, tableConfigs)
+      .catch((e) => {
+        setNodeIntializationError(true);
+        throw e;
+      })
+      .finally(() => {
+        setIsInitializingNodes(false);
+      });
+  }, [
+    selectedDataSource,
+    tablesQueryConfiguration,
+    isInitializingNodes,
+    selectedManagedTables.length,
+    initializeManagedTables,
+    nodeIntializationError,
+  ]);
+
+  useEffect(() => {
+    if (
+      !selectedDataSource ||
+      !selectedDataSource.connectorId ||
+      isInitializingParents ||
+      !!parentsIntializationError ||
+      !!Object.keys(parentsById).length ||
+      !selectedManagedTables.length
+    ) {
+      return;
+    }
+    setIsInitializingParents(true);
+    void initializeParents(
+      selectedManagedTables.map((n) => n.internalId),
+      selectedDataSource
+    )
+      .catch((e) => {
+        setParentsIntializationError(true);
+        throw e;
```

Comment:
This might lead to error never been caught, right?

Reply:
It's def not going to be caught, but it will show up in dev. 
It's client side code, so there isn't much catching to be done ? What would you suggest instead ?

Reply:
Given that you store the error in the state (with `setParentsIntializationError`), we could probably handle it gracefully, WDYT?

Reply:
error wouldn't really be actionable for the user. What would you suggest for graceful handling ?